src/handler/description.rs

```diff
@@ -0,0 +1,257 @@
+///     dptree::entry()
+///         .branch(
+///             dptree::entry()
+///                 .branch(dptree::entry().branch(dptree::filter(|| true)))
+///                 .branch(dptree::entry().chain(dptree::filter(|| false))),
+///         )
+///         .branch(dptree::entry()),
+/// );
+/// ```
+pub trait HandlerDescription: Sized + Send + Sync + 'static {
+    /// Description for [`entry`](crate::entry).
+    fn entry() -> Self;
+
+    /// Description for a user-defined handler that can do practically
+    /// everything.
+    fn user_defined() -> Self;
+
+    /// Merge descriptions to get a description for a chain handler.
+    fn merge_chain(&self, other: &Self) -> Self;
+
+    /// Merge descriptions to get a description for a branch handler.
+    fn merge_branch(&self, other: &Self) -> Self;
+
+    /// Description for [`map`](crate::map).
+    ///
+    /// ## Default implementation
+    ///
+    /// By default this returns the value from
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn map() -> Self {
+        Self::user_defined()
+    }
+
+    /// Description for [`map_async`](crate::map_async).
+    ///
+    /// ## Default implementation
+    ///
+    /// By default this returns the value from
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn map_async() -> Self {
+        Self::user_defined()
+    }
+
+    /// Description for [`filter`](crate::filter).
+    ///
+    /// ## Default implementation
+    ///
+    /// By default this returns the value from
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn filter() -> Self {
+        Self::user_defined()
+    }
+
+    /// Description for [`filter_async`](crate::filter_async).
+    ///
+    /// ## Default implementation
+    ///
+    /// By default this returns the value from
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn filter_async() -> Self {
+        Self::user_defined()
+    }
+
+    /// Description for [`filter_map`](crate::filter_map).
+    ///
+    /// ## Default implementation
+    ///
+    /// By default this returns the value from
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn filter_map() -> Self {
+        Self::user_defined()
+    }
+
+    /// Description for [`filter_map_async`](crate::filter_map_async).
+    ///
+    /// ## Default implementation
+    ///
+    /// By default this returns the value from
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn filter_map_async() -> Self {
+        Self::user_defined()
+    }
+
+    /// Description for [`endpoint`](crate::endpoint).
+    ///
+    /// ## Default implementation
+    ///
+    /// By default this returns the value from
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn endpoint() -> Self {
+        Self::user_defined()
+    }
```

Comment:
Is there any place where these functions are needed except the example `descr_locations.rs`?

Reply:
Not currently, no. But I do believe that `descr_locations.rs` is a good example. You could imagine a similar thing that literally draws the tree and allow you to click through it into source code.